File: backend/app/core/academic_fields.py
"""
Single source of truth mapping a student's academic field (chosen at
registration) to (a) which subjects they should see content for, and
(b) which entry test(s) they're implicitly prepping for.

Used by:
- RegisterRequest validation + auto-deriving interested_tests
- Study/subject filtering (see "Restrict study subjects" task)
- Future: pricing tier selection (per-test/bundle pricing)
"""

FIELD_SUBJECTS: dict[str, list[str]] = {
    "pre-engineering": ["Physics", "Chemistry", "Mathematics", "English"],
    "pre-medical": ["Biology", "Chemistry", "Physics", "English"],
    "ics-physics": ["Mathematics", "Physics", "Computer Science", "English"],
    # NOTE: "Statistics" is not yet in ALLOWED_SUBJECTS (admin.py) or ingested
    # anywhere in the content library. ICS-Stats is scoped to Math + CS +
    # English until Statistics content actually exists -- do not silently
    # promise Statistics coverage until it's real.
    "ics-stats": ["Mathematics", "Statistics", "Computer Science", "English"],
}

FIELD_TESTS: dict[str, list[str]] = {
    "pre-engineering": ["ECAT", "NET"],
    "pre-medical": ["MDCAT"],
    # NOTE: ECAT here is per explicit founder instruction. UET's actual ECAT
    # eligibility is typically Physics/Chemistry/Math (Pre-Engineering) --
    # ICS-Physics students are not usually ECAT-eligible in practice at UET
    # specifically. Confirm whether "ECAT" in this app means the literal
    # UET test or an internal content category before this reaches real users.
    "ics-physics": ["ECAT", "FAST"],
    "ics-stats": ["FAST", "NET"],
}

FIELD_LABELS: dict[str, str] = {
    "pre-engineering": "Pre-Engineering",
    "pre-medical": "Pre-Medical",
    "ics-physics": "ICS (Physics)",
    "ics-stats": "ICS (Statistics)",
}

VALID_FIELDS = set(FIELD_SUBJECTS.keys())


def subjects_for_field(field: str | None) -> list[str] | None:
    if not field:
        return None
    return FIELD_SUBJECTS.get(field)


def tests_for_field(field: str | None) -> list[str] | None:
    if not field:
        return None
    return FIELD_TESTS.get(field)

File: frontend/lib/academicFields.ts
// Mirrors backend/app/core/academic_fields.py FIELD_SUBJECTS -- keep in sync.
// Used only as a fallback for legacy accounts that registered under the old
// single-field picker and never got an explicit `subjects` list.
const FIELD_SUBJECTS: Record<string, string[]> = {
  "pre-engineering": ["Physics", "Chemistry", "Mathematics", "English"],
  "pre-medical": ["Biology", "Chemistry", "Physics", "English"],
  "ics-physics": ["Mathematics", "Physics", "Computer Science", "English"],
  "ics-stats": ["Mathematics", "Statistics", "Computer Science", "English"],
};

const ALL_SUBJECTS = ["Biology", "Chemistry", "Physics", "Mathematics", "Computer Science"];

// Returns the subjects a student should see.
// Priority order:
//   1. Explicit `subjects` (new registration flow -- student directly
//      multi-selected these at signup).
//   2. Field-derived subjects (legacy accounts from the old single-field
//      picker, before explicit subject selection existed).
//   3. ALL_SUBJECTS as a last-resort fallback so nobody sees an empty list.
// Always filtered down to ALL_SUBJECTS since English/Statistics have no
// ingested content yet -- remove that filter once that content exists.
export function subjectsForField(
  field: string | null | undefined,
  subjects?: string[] | null
): string[] {
  const explicit = subjects && subjects.length > 0 ? subjects : null;
  const derived = !explicit && field ? FIELD_SUBJECTS[field] : null;
  const list = explicit || derived || ALL_SUBJECTS;
  return list.filter((s) => ALL_SUBJECTS.includes(s));
}